Feature/implement daily bcsd

dgergel opened the pull request:
adds NASA-NEX modification to classic BCSD bias correction method with quantile mapping by day groups rather than by months, uses a new custom `PaddedDOYGrouper` to do this. To fit quantile maps by day group rather than month, the model is fit with the flag `time_grouper='daily_nasa-nex'`, e.g. 

`model_nasanex = BcsdTemperature(time_grouper='daily_nasa-nex', return_anoms=False).fit(X_train, 
                                                                                       y_train)`

This currently doesn't support leap days, that's a future to-do. 

ref #27

dgergel commented:
@jhamman - would be awesome to get a review from you on this PR when you get a chance.

jhamman commented on skdownscale/pointwise_models/utils.py:
```diff
@@ -166,3 +166,17 @@ def ensure_samples_features(obj):
+def check_datetime_index(obj, timestep):
+    """ helper function to check datetime index for compatibility
+    """
+    if isinstance(obj, pd.DataFrame):
```
what should happen when obj is not a DataFrame?

dgergel replied:
actually don't think we really need this function - I didn't end up using it in the NASA-NEX daily implementation and I think it's redundant with the testing that is in place now. seem reasonable @jhamman?

jhamman replied:
Right. I don't think we need this anymore.

dgergel replied:
Cool - just took this out.

jhamman commented on skdownscale/pointwise_models/groupers.py:
```diff
@@ -0,0 +1,49 @@
+        self.df = df
+        self.offset = offset
+        self.max = 365
+        self.days_of_year = np.arange(1, 366)
```
Which calendars will this support? I'm thinking it will be useful to check that the calendar of df.index is valid for this grouper method.

jhamman commented on skdownscale/pointwise_models/groupers.py:
```diff
@@ -0,0 +1,49 @@
+        sec_half = self.days_of_year_wrapped[self.n + self.offset : i + total_days]
+        all_days = np.concatenate((first_half, np.array([self.n]), sec_half), axis=0)
+
+        assert len(set(all_days)) == total_days, all_days
```
Let's convert this to a proper `ValueError`:

```python
if len(set(all_days)) != total_days:
    raise ValueError('...say something meaningful...')
```

jhamman commented on skdownscale/pointwise_models/groupers.py:
```diff
@@ -0,0 +1,49 @@
+        list_result = []
+        for key, group in self:
+            list_result.append(group.mean().values[0])
+        result = pd.Series(list_result, index=self.days_of_year)
```
shouldn't this be a DataFrame. In Pandas, `df.groupby(lambda x: x.month).mean() -> pd.DataFrame` if `df` is a DataFrame.

dgergel replied:
I think it should actually be a `Series`? It seems like if I make it a DataFrame, the time index gets wonky

dgergel replied:
ahh nm, got this to work in conjunction along with some other updates for your comment below

jhamman commented on skdownscale/pointwise_models/groupers.py:
```diff
@@ -0,0 +1,49 @@
+        return self.n - 1, result
+
+    def mean(self):
+        list_result = []
+        for key, group in self:
+            list_result.append(group.mean().values[0])
```
I think this will be slightly faster if you allocate a numpy array via `np.full` and populate the values as `result[key] = group.mean().values[0]`

dgergel replied:
Done, thanks!

jhamman commented:
@dgergel - anything I can do to help get this finished up?

dgergel commented:
@jhamman great timing - I'm working on these edits now and hoping to tie this up soon as well. 

For your first comment about tests, were you thinking along the lines of unit tests for the PaddedDOYGrouper and NASA-NEX option or some figures that compare it to grouping by month..or both?

jhamman commented:
> For your first comment about tests, were you thinking along the lines of unit tests for the PaddedDOYGrouper and NASA-NEX option or some figures that compare it to grouping by month..or both?

I was just thinking unit tests at this stage.

jhamman commented on skdownscale/pointwise_models/bcsd.py:
```diff
@@ -5,31 +5,56 @@
     _timestep = 'M'
 
-    def __init__(self, time_grouper=MONTH_GROUPER, return_anoms=True, qm_kwargs={}):
+    def __init__(
+        self,
+        time_grouper=MONTH_GROUPER,
+        climate_trend_grouper=DAY_GROUPER,
+        return_anoms=True,
+        **qm_kwargs):
+        
         self.time_grouper = time_grouper
+        self.climate_trend_grouper = climate_trend_grouper
+        self.climate_trend = MONTH_GROUPER
         self.return_anoms = return_anoms
         self.qm_kwargs = qm_kwargs
```
```suggestion
    def __init__(
        self,
        time_grouper=MONTH_GROUPER,
        climate_trend_grouper=DAY_GROUPER,
        return_anoms=True,
        qm_kwargs={}):
        
        self.time_grouper = time_grouper
        self.climate_trend_grouper = climate_trend_grouper
        self.climate_trend = MONTH_GROUPER
        self.return_anoms = return_anoms
        self.qm_kwargs = qm_kwargs
```

Changing this back should fix a few of your failing tests.

dgergel replied:
good catch @jhamman, thanks

dgergel commented:
@jhamman just finally got around to adding some unit tests and addressed your other comments as well. Looking forward to hearing what you think. Also note that I added in support for leap days into the `nasa-nex` option.

dgergel commented:
@jhamman I think this should be ready to merge unless you see anything else that you think needs additional mods.

jhamman commented:
Thanks @dgergel! Onward!
